sessionctx/variable/sysvar.go

```diff
@@ -1333,6 +1333,10 @@ var defaultSysVars = []*SysVar{
 		s.EnableChangeMultiSchema = TiDBOptOn(val)
 		return nil
 	}},
+	{Scope: ScopeGlobal, Name: TiDBEnableAutoIncrementInGenerated, Value: BoolToOnOff(DefTiDBEnableAutoIncrementInGenerated), Type: TypeBool, SetSession: func(s *SessionVars, val string) error {
+		s.EnableAutoIncrementInGenerated = TiDBOptOn(val)
+		return nil
+	}},
```

Comment:
There is a bug here. Because the Scope is only `ScopeGlobal` the `SetSession` function (which is also an init function) won't be called when a session starts. So the session property `s.EnableAutoIncrementInGenerated` is not set.

There are two workarounds:
- Change the Scope to Session + Global so the init function runs.
- Instead of checking the value as `ctx.GetSessionVars().EnableAutoIncrementInGenerated`, use a global var accessor with `ctx.GetSessionVars().GlobalVarsAccessor.GetGlobalSysVar(variable.TiDBEnableAutoIncrementInGenerated)`

Reply:
FYI; I've added a test for this now: https://github.com/pingcap/tidb/pull/26634

Reply:
Thanks! That explains why I couldn't get it to work...

But I'm still somewhat confused. Why does this work with a global sysvar like `tidb_enable_change_multi_schema`? I'm accessing it exactly the same way.

Reply:
Ah! Because of this: https://github.com/pingcap/tidb/blob/f4a41bd57b0c374d6da756f91c546a8039b1d132/sessionctx/variable/sysvar.go#L531-L535

Reply:
I made it session + global instead. It seems to work but looks like an entirely unrelated test is failing?

Reply:
Oh it succeeded now... flaky test perhaps?

Reply:
Yes, the list is left over from: https://github.com/pingcap/tidb/pull/24836

The PR was a refactor. The semantics are not MySQL-like, because if the global value is changed, it won't apply to active sessions immediately.. but it should.